File: solve/cpp/trotter.cc
#include "common.hh"

int d[MAX + 1];
char b[MAX + 2];

void precompute()
{
    fill(d, d + n + 1, -1); 

    b[0] = LAST;
    for (int i = 1; i <= n; ++i)
    {
        b[i] = FIRST + i - 1;
    } 
    b[n + 1] = LAST;    
}

void trotter()
{
	while (true)
	{
        output(b + 1);

		int maxi = 0; char maxv = 0;
		for (int i = 1; i <= n; ++i)
        {	
            if (b[i] > b[i + d[i]] && b[i] > maxv) 
            { 
                maxi = i;
                maxv = b[i];
            } 
        }
        if (maxi == 0) break; 

        int i = maxi + d[maxi];
		swap(b[maxi], b[i]);
		swap(d[maxi], d[i]);

		for (int i = 1; i <= n; ++i)
		{
			if (b[i] > maxv) 
            { 
                d[i] *= (-1);
            }
		}
	}
}

int main(int argc, char *argv[])
{
	if (init(argc, argv, UNORDERED) == 0)
    { 
        precompute(); 
        trotter();       
        return final();
    } 
    return 0;
}

File: solve/lua/trotter.lua
#! /usr/bin/lua

require "common"

d = {}

function trotter()
    while true do
        output()

        local k = n + 2
        for i = 1, n do
            if a[i] > a[i + d[i]] and a[i] > a[k] then
                k = i
            end
        end
        if k == n + 2 then
            break
        end

        local j = a[k]
        local k2 = k + d[k]
        a[k], a[k2] = a[k2], a[k]
        d[k], d[k2] = d[k2], d[k]
        for i = 1, n do
            if a[i] > j then
                d[i] = d[i] * (-1)
            end
        end
    end
end

skeleton(UNORDERED, function ()
    a[0] = n + 97
    a[n + 1] = n + 97
    a[n + 2] = 0

    for i = 1, n do
        d[i] = -1
    end

    trotter()
end)
